WP 388 config organization

mmcgahan opened the pull request:
Fixes https://meetup.atlassian.net/browse/WP-388

Because the different CLI commands need to share config information, and we can't fully anticipate which config will need to be shared in the future, I think it makes most sense to consolidate all 'config' into a single directory near the root of the CLI and essentially treat it as a single 'app config' module with no dependencies. Within the `config` directory, we can group related config together by script or directory, but the public interface for all config will be a single `config/index.js` module.

_There is no new functionality in this branch_ - it just an organization of exported objects, import paths, and related variable names.

The one follow-up that will need to happen in the meetup web platform is to change [this import statement](https://github.com/meetup/meetup-web-platform/blob/e91c321b699334651dacba60ba449f8b2bd84e07/src/plugins/service-worker/index.js#L2) to point to the right config location

mattkime commented on src/commands/txCommands/util/index.js:
```diff
@@ -29,9 +29,9 @@ const checkEnvVars = () => {
-	Rx.Observable
-		.bindNodeCallback(fs.readFile)(filepath)
-		.map(buffer => buffer.toString());
+	Rx.Observable.bindNodeCallback(fs.readFile)(filepath).map(buffer =>
```
formatting changes make this diff much more difficult to read. either prettier should be integrated into the repo or formatting changes should be left alone.

mattkime commented on src/commands/buildCommands/trn.js:
```diff
@@ -36,31 +38,33 @@ const writeTrnModules = messagesByLocale => ({ filename, msgids }) => {
  *   react-intl babel plugin output for each component that calls `defineMessages`
  */
-const buildTrnModules = localeCodes =>
+const buildTrnModules = () =>
```
good catch

mattkime commented:
I think its worth redefining 'config' for this project.

Starting with the obvious, config is info that customizes the behavior of code. Config for mwp-cli lives in the consumer repo's `project.json`. The items that are being reorganized here are mostly config for utilities used by individual commands. To webpack it may be config, but within the mwp-cli repo its code. Thinking about it this way, we should have a place for shared code but everything else should be contained in the appropriate command directory.

mmcgahan commented:
@mattkime Yeah, 'config' covers a huge range of information about a project, and it's hard to know ahead of time whether a particular 'config' value is exclusively used by a particular command/repo/invocation, which is why I think consolidating the config closer to the root makes sense. I agree in principle that config should live as close as possible to the code that consumes it, but the config for mwp-cli has a very broad application and we're already finding cases where we want to consume config that was initially defined elsewhere.

> Config for mwp-cli lives in the consumer repo's `project.json`

Unfortunately it's much more complicated - I've found it helpful to think about config having many different layers with different lifetimes

1. platform-general config (all apps), e.g. `supportedLocales` and babel presets/plugins
2. app-specific config, e.g. `package.json` `config.subdomain`
3. env-specific config - e.g. dev-vs-prod-vs-test
4. CLI command-specific config (build vs run)
5. CLI option-specific config (--locales)
6. module-specific config (not really applicable to this repo, but e.g. config for Hapi plugins in MWP)

This makes for sprawling config combinations that can't be all be hard-coded into `package.json`, which only really make sense for 'level 2' config.

Each of these layers could import config from a different location, but many/most of the config ends up being used in combinations that are difficult to predict ahead of time. For example, many of the `paths` appear to be specific to webpack, and therefore the `build` command, but we also need to know about webpack config for the `run` command - would that config live at the root? `env` info is currently only used by `build`, but it is general environment information so in theory it could be used in many different places - should it be in `build`, or in a more generic location? The babel config info is currently only used inside webpack config, but babel can be run independently (which I'm working on in #10), so should the babel config be put inside `webpack` or somewhere higher up? We also use babel for transifex scripts, so should we split the babel config files between the two commands and try to keep them in sync manually?

> The items that are being reorganized here are mostly config for utilities used by individual commands

Some of the config is specific to particular command utilities, but I'm not sure that it's the majority - e.g.`paths`, `locales`, and `package` are used in more than one command, and something like `env` is not command-specific. Some of this config is also usable as imports from projects that depend on this repo.

Having worked against some of the config challenges in MWP, I've found that splitting out config files in order to keep it 'organized' can make usage very awkward (e.g. 3+ `import` statements to collect the required config info for a particular module). The solution I've implemented here is to consolidate _all_ this kind of config into a central module (`config/index.js`) that provides a single, structured interface to 'config' with no internal dependencies.

mattkime commented:
>it's hard to know ahead of time whether a particular 'config' value is exclusively used by a particular command/repo/invocation

I think the best way to deal with this is to code for present needs rather than worrying about an undefined future.

> For example, many of the paths appear to be specific to webpack, and therefore the build command, but we also need to know about webpack config for the run command - would that config live at the root?

The bit that is shared would live at root. The majority (80%?) of paths are used exclusively by the build utils.

Overall I'm unconvinced that this is an improvement. Maybe we need to google hangout about it.

mmcgahan commented:
Prettier applied to `master` and diff updated

mmcgahan commented:
Just to try to lay out the cost/benefits of this reorganization more clearly:

Pro:
- central location to put (all?) platform-related config
- clear public interface with config (`src/config/index.js`)
- stable public interface

Con:
- config that is currently used in one place is not co-located (e.g. webpack `rules`)

An alternative approach would be to co-locate config as much as possible, but move config to 'higher levels' in this repo as it gets shared with more modules, e.g. webpack `rules` would be in `src/commands/buildCommands`, webpack `paths` would be in `src/` because they are used by multiple commands.

Pro:
- config is located close to where it is consumed
- 'private' config is more difficult to discover

Con:
- when config needs change, the config modules need to move, which means `require`/`import` paths need to be changed in unrelated code - perhaps even in other repos
- config for a specific feature may be spread among many different config modules
- potentially many similar-but-different config modules, leading to ambiguity, e.g. `src/config/paths.js` for shared path definitions, `src/commands/buildCommands/config/paths.js` for build-specific paths, `src/commands/config/paths.js` for build + run paths

---

The latter approach is definitely more in line with 'code for present needs', but my experience with that approach in other platform projects leads me to believe that avoiding the effort to make the architecture adaptable to future needs will result in significant maintenance burden.

---

Areas where this PR makes a difference:

- [`src/commands/runScripts/_webpack-dev-server.js`](https://github.com/meetup/mwp-cli/pull/9/files#diff-fb3e897cb83b7ce21f6c7f4f62dc2b4dL5) was importing from `'../buildCommands/config'`
- [`src/commands/runScripts/start-dev.js`](https://github.com/meetup/mwp-cli/pull/9/files#diff-5ce647d8e2c2e9b6c823d6c5f0320157L12) was importing from `'../buildCommands/config'`
- [`src/commands/txCommands/util/index.js`](https://github.com/meetup/mwp-cli/pull/9/files#diff-b7ed5127b5fcf4b7bfb5d049d728fc07L6) was importing from `'../../buildCommands/config/paths'` and `'../../../util/packageConfig'` (now a single config import)
- [`src/util/packageConfig.js`](https://github.com/meetup/mwp-cli/pull/9/files#diff-a5da13c4a8a7fe37840ce522f8dc8aa8L2) was importing from `'../commands/buildCommands/config/paths'`
- [`src/util/addLocalesOption.js`](https://github.com/meetup/mwp-cli/pull/9/files#diff-c967779de6ebd0847d0ded1233c6c4c3L1) was importing from `'../commands/buildCommands/config/locales'`

Each of these 'distant' config imports now point to the same `src/config/index.js` file - the config interface is much more consistent and therefore predictable. Breaking out the config into smaller modules throughout the `commands/` dir feels like a step backward IMO.

mattkime commented:
I've come around to the idea of centralizing config  although i still think `paths` is too much of a grab bag but i'm not going to hold things up over it.

mmcgahan commented:
@mattkime I was thinking more about `paths` and I'm wondering if we might just pare it down a bit in a followup to include mainly to some 'standard' directories (src, build, maybe 2-3 other nested dirs) - then it would be up to the importing modules (e.g. the webpack config modules) to use those 'base path' references and build their own module-specific paths. I think that might make `paths` more focused without creating sprawl elsewhere

I also want to add a quick README to this branch before I merge just to keep the overview of the config organization accessible

mattkime commented:
just tossing out an idea - maybe `run` and `build` are uniquely joined in such a way that additional commands won't be since they both heavily use webpack.

mattkime commented:
...and another idea..... in our conversation yesterday you mentioned having a concern with where some config lived so it could be accessible to other projects. such exports should be explicitly defined. i don't think this should determine where the code exists inside the project...just provide a layer of abstraction so code can be refactored without concern.
